### PR TITLE
Feat/augmentation

### PR DESCRIPTION
**Pull request by edgarriba**
This PR adds an initial proposal for data augmentation

depends off https://github.com/PyTorchLightning/lightning-Covid19/pull/17

### CODE REVIEW

**Comment by oplatek** [comment]
Hi @edgarriba, can you pls fix the pep8/flake8 errors?

File: lightning_covid19/model/densenet.py [hunk 1]
```diff
@@ -167,12 +195,26 @@ def prepare_data(self):
         else:
             print('Cloning the covid chestxray dataset. It may take a while\n...\n', flush=True)
             run(f'git clone {clone_uri} {chestxray_root}')
+
         transform = transforms.Compose([xrv.datasets.XRayCenterCrop(),
                                         xrv.datasets.XRayResizer(self.hparams.xraysize)])
+
```
Comment (oplatek) [inline lightning_covid19/model/densenet.py]: The variable transform is left unused

File: lightning_covid19/model/densenet.py [hunk 2]
```diff
@@ -12,13 +12,38 @@
...
+import kornia as K
 import torchvision.transforms as transforms
 import torchxrayvision as xrv
-from pytorch_lightning.core import LightningModule
 
+from pytorch_lightning.core import LightningModule
 from lightning_covid19.utils import run
 
 
+class KorniaAugmentation(nn.Module):
```
Comment (oplatek) [inline lightning_covid19/model/densenet.py]: Can we rename it to something descriptive? I.e. what the Augmentation does.

You may mention that it uses Kornia augmentation in the docs string.

File: lightning_covid19/model/densenet.py [hunk 3]
```diff
@@ -167,12 +195,26 @@ def prepare_data(self):
         else:
             print('Cloning the covid chestxray dataset. It may take a while\n...\n', flush=True)
             run(f'git clone {clone_uri} {chestxray_root}')
+
         transform = transforms.Compose([xrv.datasets.XRayCenterCrop(),
                                         xrv.datasets.XRayResizer(self.hparams.xraysize)])
+
+        # TODO: we could just use K.augmentation.RandomResizedCrop
+        # https://kornia.readthedocs.io/en/latest/augmentation.html#kornia.augmentation.RandomResizedCrop
```
Comment (oplatek) [inline lightning_covid19/model/densenet.py]: Why don't we use the TODO right now?

**Review by oplatek — COMMENTED** [review]
Can you also pls update your branch with the merged #17. There are conflicts. Thanks!

File: lightning_covid19/model/densenet.py (hunk 1 above)
Comment (edgarriba) [reply lightning_covid19/model/densenet.py]: this was just to replicate their pre processing with kornia. Depending what strategy we follow I could internally send a PR to torchxrayvision and implement kornia there. This two transformations I think are very specific to the dataset. The first, center crops based on the half of the minimum H/W of the image, the second calls resize from skimage to normalise for batching.

File: lightning_covid19/model/densenet.py (hunk 2 above)
Comment (edgarriba) [reply lightning_covid19/model/densenet.py]: sure, but it's just a generic class to encapsulate the augmentations pipeline to make it easy for anyone to play with different transforms. What about `DataAugmentatorPipe` ?

File: lightning_covid19/model/densenet.py (hunk 3 above)
Comment (edgarriba) [reply lightning_covid19/model/densenet.py]: I think it needs some discussion on the strategy about how to use the data.
just to show a couple random cherry picked examples:

image

image

image

not sure what would be the best. My guess would the best to analyses the whole image and create some kind of attention to not miss any part.

Another technical note: images are in png, jpg, jpeg as far I can appreciate.

File: lightning_covid19/model/densenet.py (hunk 1 above)
Comment (oplatek) [reply lightning_covid19/model/densenet.py]: I agree - let's move to kornia - no problem here.
The transform variable on line is just left unused... let's delete it!

File: lightning_covid19/model/densenet.py (hunk 2 above)
Comment (oplatek) [reply lightning_covid19/model/densenet.py]: Maybe I am wrong but right now it is not a generic class. It is a particular instance which applies

- RandomHorizontalFlip
- ColorNormalizatoin
- ColorJitter

I see it as a nice example how to use Kornia data augmentation functionalities but not a "API defining" class. Am I wrong?

Comment (oplatek) [reply lightning_covid19/model/densenet.py]: My suggestion is to name it

```
ColorHFlipAugmentation(nn.Module):
    """"An example how to use Kornia data augmentation running RandomHorizontalFlip, Color.Normalize and color jitter""""
```

I want to get this right (and understand it properly) - so it does not create confusion later.

Comment (oplatek) [reply lightning_covid19/model/densenet.py]: More fitting name is welcome - the naming matters :)

File: lightning_covid19/model/densenet.py (hunk 3 above)
Comment (oplatek) [reply lightning_covid19/model/densenet.py]: > My guess would the best to analyses the whole image and create some kind of attention to not miss any part.

I absolutely agree.
I created an issue describing one possible solution using Bounding Boxes in an issue: #20

Can you please convert this TODO to the Github issue - add it to the #20 issue, create a more general issue, or suggest your favorite solution in a new issue.

**Main point: please remove the TODO from the code itself. :)** 

Thanks!

File: lightning_covid19/model/densenet.py (hunk 2 above)
Comment (Borda) [reply lightning_covid19/model/densenet.py]: I guess that the applyed transformations are parameters that can be optimaly set on/off, right? Then I would keep the name a bit general as @edgarriba proposed...

File: lightning_covid19/model/densenet.py (hunk 3 above)
Comment (Borda) [reply lightning_covid19/model/densenet.py]: If you want, there is bot converting todo to issues automatically lol

File: lightning_covid19/model/densenet.py (hunk 2 above)
Comment (edgarriba) [reply lightning_covid19/model/densenet.py]: @Borda right. Since the project will need different trial and error experiments I think it's good to have this generic class were each of us can enable/disable/add/remove augmentation layers.

File: lightning_covid19/model/densenet.py (hunk 3 above)
Comment (edgarriba) [reply lightning_covid19/model/densenet.py]: @oplatek will comment there. @Borda would love to see that bot xD

Comment (Borda) [reply lightning_covid19/model/densenet.py]: https://github.com/marketplace/todo shall I add it?

Comment (Borda) [reply lightning_covid19/model/densenet.py]: seems not for python :{

File: lightning_covid19/model/densenet.py (hunk 2 above)
Comment (oplatek) [reply lightning_covid19/model/densenet.py]: oh, I understand now

**Comment by oplatek** [comment]
@edgarriba LGTM  -> I will merge

Note: the Travis was not passing except because of this minor problem https://travis-ci.org/github/PyTorchLightning/lightning-Covid19/jobs/668325985#L329
@edgarriba do not worry about it, I will fix it in the master branch.

@Borda questions about ^: How can I collaborate on MR with the author and help him fixing such simple stuff if I cannot push to `edgarriba:feat/augmentation`?

**Comment by edgarriba** [comment]
@oplatek you can rebase author PR, modify and then push/merge

**Comment by Borda** [comment]
In PR you can use suggestions which can be simply merged from web UI or edit comple file which is part of the PR also in web UI =)
